IPython/terminal/prompts.py

```diff
@@ -77,3 +83,52 @@ def write_output_prompt(self):
                 self.shell.pt_cli.print_tokens(tokens)
             else:
                 sys.stdout.write(prompt_txt)
+
+    def write_format_data(self, format_dict, md_dict=None):
+        """Write the format data dict to the frontend.
+
+        This version adds support for image/png if the 
+        IPYTHON_IMAGE_VIEWER environment avariable is set.
+
+        Parameters
+        ----------
+        format_dict : dict
+            The format dict for the object passed to `sys.displayhook`.
+        md_dict : dict (optional)
+            The metadata dict to be associated with the display data.
+        """
+        image_data = format_dict.get("image/png")
+        if image_data is not None and self.view_image(image_data):
+            return
+        return super().write_format_data(format_dict, md_dict)
+
+    def view_image(self, image_data):
+        """Attempt to display the raw image_data .
+          image_data should be a `bytes` object containing image data in PNG format
+
+          We will check if the IPYTHON_IMAGE_VIEWER environment variable is set.
+          If so, this is used to show the image.
+        """
+        # Probably not worth caching the image_viewer.
+        # Not caching has the advantage that you can set it in a running IPython.
+        image_viewer = os.getenv("IPYTHON_IMAGE_VIEWER")
+        if image_viewer is None:
+            # Environment variable does not exist
+            return False
+        # We use shlex.split so that IPYTHON_IMAGE_VIEWER can also contain
+        # options in addition to the executable.
+        image_viewer = shlex.split(image_viewer)
+        if not image_viewer:
+            # Environment variable existed but was empty
+            return False
+        with tempfile.TemporaryDirectory() as dirname:
+            png_file = os.path.join(dirname, "out.png")
+            with open(png_file, "wb") as f:
+                f.write(image_data)
+            try:
+                # We don't want to clutter the display with errors on failure,
+                # so redirect stderr to DEVNULL.
+                subprocess.check_call(tuple(image_viewer) + (png_file,), stderr=subprocess.DEVNULL)
```

Comment:
ok, so that works with blocking image viewers, not with non blocking one as the temporary directory will be torn down before the viewer actually read the the file. We'll have to (likely) attach the teardown to a sys.exit hook, or to the __del__ method of `RichPromptDisplayHook`. Maybe with a cache to not let more than a couple of things open. 
I'm also thinking that the viewer should likely not be a subprocess, but a python callable, so that you __can__ hook into pure-python display system like for example inline images. 

We'll have to also figure out a clean way to tell matplotlib it can use inline.